### ISSUE TITLE
Analysis fail to complete

### ISSUE THREAD

**Issue by idyuham** [inasafe/inasafe#4304]
### Problem

I was trying to run analysis of floods on roads, then building then on pop, all of the failed. I run them separately.  UGANDA dateset provided for training in Kampala.
```
In file "C:/Users/STEVE SAPIN/.qgis2/python/plugins\inasafe\safe\impact_function\impact_function.py", line 1135, in run self._run()
In file "C:/Users/STEVE SAPIN/.qgis2/python/plugins\inasafe\safe\utilities\profiling.py", line 122, in with_profiling ret = fn(*args, **kwargs)
In file "C:/Users/STEVE SAPIN/.qgis2/python/plugins\inasafe\safe\impact_function\impact_function.py", line 1296, in _run self.gis_overlay_analysis()
In file "C:/Users/STEVE SAPIN/.qgis2/python/plugins\inasafe\safe\utilities\profiling.py", line 122, in with_profiling ret = fn(*args, **kwargs)
In file "C:/Users/STEVE SAPIN/.qgis2/python/plugins\inasafe\safe\impact_function\impact_function.py", line 1417, in gis_overlay_analysis self.aggregate_hazard_preparation()
In file "C:/Users/STEVE SAPIN/.qgis2/python/plugins\inasafe\safe\utilities\profiling.py", line 122, in with_profiling ret = fn(*args, **kwargs)
In file "C:/Users/STEVE SAPIN/.qgis2/python/plugins\inasafe\safe\impact_function\impact_function.py", line 1806, in aggregate_hazard_preparation self._aggregate_hazard_impacted = union(self.hazard, self.aggregation)
In file "C:/Users/STEVE SAPIN/.qgis2/python/plugins\inasafe\safe\utilities\profiling.py", line 122, in with_profiling ret = fn(*args, **kwargs)
In file "C:/Users/STEVE SAPIN/.qgis2/python/plugins\inasafe\safe\gis\vector\union.py", line 101, in union index_b = create_spatial_index(union_a)
In file "C:/Users/STEVE SAPIN/.qgis2/python/plugins\inasafe\safe\utilities\profiling.py", line 122, in with_profiling ret = fn(*args, **kwargs)
In file "C:/Users/STEVE SAPIN/.qgis2/python/plugins\inasafe\safe\gis\vector\tools.py", line 251, in create_spatial_index spatial_index = QgsSpatialIndex(layer.getFeatures())
```

### Environment
* InaSAFE : 
* QGIS : 
* OS : 
* Data used (filenames and paths so as to reproduce, a project is better):
  * Exposure : 
  * Hazard : 
  * Aggregation :
  * Extent : 
  
Checklist (if applicable): 
* [ ] Did you try to run it in debug mode ?
* [ ] Is the data available ?
  * If yes, where (BTSync, data.inasafe.org, zip) ?

**Comment by Gustry** [comment]
Thanks for reporting the issue.
This is due to invalid geometries. I'm try to have a look.

**Comment by Gustry** [comment]
I'm trying to replicate, but I can't.

This exception is raised because of invalid geometries when we create the spatial index. But we are already doing a first step to validate geometries (before the union algo starts), so I don't know what else I can do for now (as I can't replicate)

**Comment by ismailsunni** [comment]
Perhaps it's OS specific? We can ask @samnawi or @felix-yew to check

**Comment by Gustry** [comment]
She was on MacOS ;-)

**Comment by ismailsunni** [comment]
Really? The path looks like a Windows path.

**Comment by Gustry** [comment]
Ah yes sorry. But I got one woman with the same error in Kampala, she was using MacOs.

**Comment by Gustry** [comment]
I fixed another ticket related to invalid geometries. I don't know if my PR could fix this one. I can't replicate this issue so I'm updating labels.

**Comment by Charlotte-Morgan** [comment]
@felix-yew  - could you take a look at this and see if you can get it to run. Kartoza will share the data with you

**Comment by timlinux** [comment]
@timlinux to put the Uganda data into btsync for @felix-yew to try to replicate

**Comment by felix-yew** [comment]
Hi @Charlotte-Morgan @timlinux - sorry for the delay response, where I can find the Uganda data? I can not find in `Real World Data`

**Comment by Charlotte-Morgan** [comment]
hi @felix-yew - I think maybe @timlinux has not shared the data yet. maybe @Gustry has the data he can share?

**Comment by felix-yew** [comment]
Hi @Gustry - could you please to share the data? so I can try this issue. Thank you.

**Comment by timlinux** [comment]
@timlinux to get the data to @felix-yew

**Comment by timlinux** [comment]
To do as bug fix post 4.2 release

**Comment by timlinux** [comment]
@felix-yew the RO BTsync key is BWHM2A7KLHG5LFARSOZMQ4KUL7AM25XM6 (includes the full training course memory stick contents)

**Comment by felix-yew** [comment]
Ok thank you @timlinux - I already add it. It maybe will sync later.

**Comment by Charlotte-Morgan** [comment]
hi @felix-yew if you have the data now, maybe you can see if you can replicate this problem

**Comment by felix-yew** [comment]
Hi @idyuham - I want to know what the data that you use to run all of the impact function, for the flood data and the exposure data. Because, there are a lot of data that I see in the folder of `GIS_Data`. Could you please to show me where is the data that you use? Because you not mention the detail of the data that you use in the ticket. 

Cc: @timlinux @Charlotte-Morgan

**Comment by Gustry** [comment]
As I said, this ticket is difficult to replicate because it's due to invalid geometries. It might be very specific to the platform. We got this issue during the training on only 1 computer at the same time (IIRC)

**Comment by Gustry** [comment]
Just to clarify quickly, layers are valid at the beginning of the impact function (the exposure, the hazard and the aggregation), but after some geoprocessing, some layers might become invalid. Typically, the union, intersection, or the rasterize algorithms have a probability to produce invalid geometries. In InaSAFE, we have some process to clean them and also QGIS will skip invalid geometries. But this exception I see in this issue is coming from the index creation (`QgsSpatialIndex(layer.getFeatures())`). I already put before this step an buffer(0) to clean geometries, but it seems it's not enough. An exception will be raised in the QgsSpatialIndex.

To sum up, these invalid geometries are created during the analysis itself and depend on the plateform I guess because we can't replicate the issue on another computer.

IIRC about the dataset, I would say in `GIS_Data` folder:
* Uganda_Flood_2007
* Uganda_OSM_Roads
* Uganda_OSM_Buildings
* Uganda_Population

**Comment by felix-yew** [comment]
Hi @idyuham @Gustry - I try to run flood on road, I use:
uganda_flood_2007 --> hazard
osm_roads --> `Uganda_OSM_roads` for exposure
I run it with InaSAFE 4.2.3 it can work properly
![image](https://user-images.githubusercontent.com/13621886/31433222-d616b4d8-aea2-11e7-9a38-449af0543446.png)

**Comment by Gustry** [comment]
Sorry, as we can't replicate the issue, I'm going to close this ticket.

**Comment by Gustry** [comment]
I could replicate the issue with one dataset provided by Tim today

**Comment by wonder-sk** [comment]
This PR fixes an unhandled exception that may be raised from QgsSpatialIndex constructor: https://github.com/qgis/QGIS/pull/6953 - still need to verify whether it is actually causing this issue or it is unrelated...

**Comment by Gustry** [comment]
Hi @wonder-sk 
Thanks for your patch in QGIS.

I doubt that layers were empty when we got this exception.

Sorry I warned you, it's difficult to replicate this issue. I couldn't really replicate this issue 100%.
Maybe it would be good to wait on the porting to QGIS 3 and check with the new architecture on QGIS 3.2? @timlinux ?
